### PR TITLE
feat: ttid

### PR DESCRIPTION
**Pull request by buenaflor**
## :scroll: Description

Implement TTID 

## :bulb: Motivation and Context

Split pr of #1882 

## :green_heart: How did you test it?
CI, manual test with spotlight

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPii` is enabled
- [x] I updated the docs if needed
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: flutter/lib/src/sentry_flutter_measurement.dart [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import '../sentry_flutter.dart';
+
+extension SentryFlutterMeasurement on SentryMeasurement {
+  /// Duration of the time to initial display in milliseconds
+  static SentryMeasurement timeToInitialDisplay(Duration duration) {
```
Comment (stefanosiano) [inline flutter/lib/src/sentry_flutter_measurement.dart]: you can create new constructors for both ttid and ttfd in `SentryMeasurement` without the need of extensions
if you want to keep the assertion on the duration you can even create a `factory` constructor (still in `SentryMeasurement`)

Comment (buenaflor) [reply flutter/lib/src/sentry_flutter_measurement.dart]: I thought it'd be nice to separate it since `SentryMeasurement` is in the `dart` package and ttid/ttfd don't exist in only dart

Comment (stefanosiano) [reply flutter/lib/src/sentry_flutter_measurement.dart]: we already have frames stuff in `SentryMeasurement`, that are not used in pure dart apps, so it's fine to put everything in one place

File: flutter/lib/src/navigation/time_to_display_transaction_handler.dart [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  TimeToDisplayTransactionHandler({
+    required Hub? hub,
+    required bool? enableAutoTransactions,
+    required Duration? autoFinishAfter,
+  })  : _hub = hub ?? HubAdapter(),
+        _enableAutoTransactions = enableAutoTransactions,
+        _autoFinishAfter = autoFinishAfter,
+        _native = SentryFlutter.native;
+
+  Future<ISentrySpan?> startTransaction(String? routeName, Object? arguments,
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/time_to_display_transaction_handler.dart]: we may move this method inside `TimeToDisplayTracker` and remove this class, as it's only used there

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 3]
```diff
@@ -179,77 +214,14 @@ class SentryNavigatorObserver extends RouteObserver<PageRoute<dynamic>> {
...
-  Future<void> _finishTransaction() async {
-    final transaction = _transaction;
-    _transaction = null;
-    if (transaction == null || transaction.finished) {
-      return;
-    }
-    transaction.status ??= SpanStatus.ok();
-    await transaction.finish();
+    await _timeToDisplayTracker?.startTracking(routeName, arguments);
+    completedDisplayTracking?.complete();
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]:
```suggestion
    _completedDisplayTracking?.complete();
```

File: flutter/lib/src/navigation/time_to_display_tracker.dart [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  final bool _enableTimeToFullDisplayTracing;
+
+  TimeToDisplayTracker({
+    required bool enableTimeToFullDisplayTracing,
+    required TimeToDisplayTransactionHandler ttdTransactionHandler,
+    TimeToInitialDisplayTracker? ttidTracker,
+  })  : _native = SentryFlutter.native,
+        _enableTimeToFullDisplayTracing = enableTimeToFullDisplayTracing,
+        _ttdTransactionHandler = ttdTransactionHandler,
+        _ttidTracker = ttidTracker ?? TimeToInitialDisplayTracker();
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/time_to_display_tracker.dart]: we should always set the `isManual` to false when a route is pushed.
If the user use the `SentryDisplayWidget` only in one route, the others will never complete, as we use a singleton that thinks all routes are manual.
If we do it here in `startTracking`, let's make sure the manual API works as expected

Comment (buenaflor) [reply flutter/lib/src/navigation/time_to_display_tracker.dart]: I'm setting `isManual = false` by calling the `clear()` method. this happens always before the tracking starts via `_finishTransaction` https://github.com/getsentry/sentry-dart/blob/4dedf37f828d3ea4f5bf3cf7c4a95683379c2492/flutter/lib/src/navigation/sentry_navigator_observer.dart#L142

File: flutter/lib/src/navigation/time_to_initial_display_tracker.dart [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  bool _isManual = false;
+  Completer<DateTime>? _trackingCompleter;
+  DateTime? _endTimestamp;
+
+  /// This endTimestamp is needed in the [TimeToFullDisplayTracker] class
+  @internal
+  DateTime? get endTimestamp => _endTimestamp;
+
+  Future<void> trackRegularRoute(ISentrySpan transaction,
+      DateTime startTimestamp, String routeName) async {
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/time_to_initial_display_tracker.dart]: you can add a nullable `endTimestamp` and a nullable `origin` parameters, null on regular routes, and have the `trackAppStart` just call this method, as they are nearly identical.
Or you can create a new private method called by both

File: flutter/lib/src/navigation/time_to_display_tracker.dart (hunk 4 above)
Comment (stefanosiano) [reply flutter/lib/src/navigation/time_to_display_tracker.dart]: yeah, just noticed it, forget about it 😄

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,16 @@
 
 ### Features
 
+- Add TTID, allows you to measure the time it takes to render the first frame of your screen ([#1910](https://github.com/getsentry/sentry-dart/pull/1910))
```
Comment (philipphofmann) [inline CHANGELOG.md]: `l`:
```suggestion
- Add TTID (time to initial display), which allows you to measure the time it takes to render the first frame of your screen ([#1910](https://github.com/getsentry/sentry-dart/pull/1910))
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,16 @@
...
+  - Requires using the [routing instrumentation](https://docs.sentry.io/platforms/flutter/integrations/routing-instrumentation/).
+  - Introduces two modes: `automatic` and `manual`.
```
Comment (philipphofmann) [inline CHANGELOG.md]: `l`:
```suggestion
  - Introduces two modes: 
```

File: dart/lib/src/sentry_measurement.dart [hunk 8]
```diff
@@ -39,6 +39,20 @@ class SentryMeasurement {
...
+        name = 'time_to_initial_display',
+        value = duration.inMilliseconds,
+        unit = DurationSentryMeasurementUnit.milliSecond;
+
+  /// Duration of the time to full display in milliseconds
+  SentryMeasurement.timeToFullDisplay(Duration duration)
+      : assert(!duration.isNegative),
+        name = 'time_to_full_display',
+        value = duration.inMilliseconds,
+        unit = DurationSentryMeasurementUnit.milliSecond;
```
Comment (philipphofmann) [inline dart/lib/src/sentry_measurement.dart]: `l`: Isn't this PR only about TTID?
```suggestion
```

File: dart/lib/src/sentry_tracer.dart [hunk 9]
```diff
@@ -109,18 +109,24 @@ class SentryTracer extends ISentrySpan {
...
-            _rootEndTimestamp = oldestChildEndTimestamp;
+        DateTime? latestEndTime;
+
+        for (var child in children) {
+          final childEndTimestamp = child.endTimestamp;
+          if (childEndTimestamp != null) {
+            if (latestEndTime == null ||
+                childEndTimestamp.isAfter(latestEndTime)) {
+              latestEndTime = child.endTimestamp;
+            }
```
Comment (philipphofmann) [inline dart/lib/src/sentry_tracer.dart]: `l`: Out of curiosity, why did you change this?

File: flutter/lib/src/navigation/sentry_display_widget.dart [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+/// ```
+///
+/// Make sure to configure Sentry's routing instrumentation in your app by following
+/// the guidelines provided in Sentry's documentation for Flutter integrations:
+/// https://docs.sentry.io/platforms/flutter/integrations/routing-instrumentation/
+///
+/// See also:
+/// - [Sentry's documentation on Flutter integrations](https://docs.sentry.io/platforms/flutter/)
+///   for more information on how to integrate Sentry into your Flutter application.
+class SentryDisplayWidget extends StatefulWidget {
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/sentry_display_widget.dart:30]: `m`: This small feature could be in an extra PR. Can you split it if it's not too much work?

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 11]
```diff
@@ -44,57 +52,66 @@ typedef AdditionalInfoExtractor = Map<String, dynamic>? Function(
...
         _setRouteNameAsTransaction = setRouteNameAsTransaction,
         _routeNameExtractor = routeNameExtractor,
-        _additionalInfoProvider = additionalInfoProvider,
-        _native = SentryFlutter.native {
+        _additionalInfoProvider = additionalInfoProvider {
     if (enableAutoTransactions) {
-      // ignore: invalid_use_of_internal_member
       _hub.options.sdk.addIntegration('UINavigationTracing');
     }
+    _timeToDisplayTracker = timeToDisplayTracker ??
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: `h`: I don't fully understand how the autoFinishAfter now works if you pass in a `timeToDisplayTracker`. How can we ensure that the `TimeToDisplayTracker` passed here uses the correct parameters for `autoFinishAfter` and `enableAutoTransactions`?

File: flutter/lib/src/navigation/time_to_display_tracker.dart [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    final transactionContext = SentryTransactionContext(
+      routeName,
+      SentrySpanOperations.uiLoad,
+      transactionNameSource: SentryTransactionNameSource.component,
+      origin: SentryTraceOrigins.autoNavigationRouteObserver,
+    );
+
+    final transaction = _hub?.startTransactionWithContext(
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/time_to_display_tracker.dart]: `m`: It seems a bit weird to me that the time to display tracker now starts the transaction. I think we should leave this responsibility to the navigation observer. Why did you choose this approach?

**Review by philipphofmann — COMMENTED** [review]
First pass.

File: flutter/lib/src/navigation/time_to_initial_display_tracker.dart [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+            ? SentryTraceOrigins.manualUiTimeToDisplay
+            : SentryTraceOrigins.autoUiTimeToDisplay);
+
+    final duration = Duration(
+        milliseconds: _endTimestamp.difference(startTimestamp).inMilliseconds);
+    final ttidMeasurement = SentryMeasurement.timeToInitialDisplay(duration);
+
+    transaction.setMeasurement(ttidMeasurement.name, ttidMeasurement.value,
+        unit: ttidMeasurement.unit);
+    await ttidSpan.finish(endTimestamp: _endTimestamp);
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/time_to_initial_display_tracker.dart:90]: why did you remove the `clear()` call?

Comment (buenaflor) [reply flutter/lib/src/navigation/time_to_initial_display_tracker.dart:90]: since we call it in `_finishTransaction` anyway in the navigation observer we are calling it twice

File: dart/lib/src/sentry_tracer.dart (hunk 9 above)
Comment (buenaflor) [reply dart/lib/src/sentry_tracer.dart]: this is basically a bug fix, will create another pr for this, this is only relevant for ttfd anyway

File: flutter/lib/src/navigation/time_to_initial_display_tracker.dart (hunk 13 above)
Comment (buenaflor) [reply flutter/lib/src/navigation/time_to_initial_display_tracker.dart:90]: unless you think we should be more defensive here and clear regardless?

Comment (stefanosiano) [reply flutter/lib/src/navigation/time_to_initial_display_tracker.dart:90]: I'd be for clearing here, so the logic behind ttid stays in the ttid tracker and not spreadto other classes

Comment (buenaflor) [reply flutter/lib/src/navigation/time_to_initial_display_tracker.dart:95]: should we maybe clear at the start and at the end? edge case, because it's also possible that users route to another screen before ttid has finished

**Comment by buenaflor** [comment]
~don't review yet, still need to fix the start timestamps~

ready for review

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 14]
```diff
@@ -242,14 +274,41 @@ class SentryNavigatorObserver extends RouteObserver<PageRoute<dynamic>> {
     await _native?.beginNativeFramesCollection();
   }
 
-  Future<void> _finishTransaction() async {
+  Future<void> _startTimeToDisplayTracking(Route<dynamic>? route) async {
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart:268]: `l`: You could rename this to `_startTransaction` and merge the current content of `_startTransaction` into this function. Now in `didPush` the code reads like this:
```
_finishTransaction();
_startTimeToDisplayTracking(route);
```

Which is a bit confusing because you finish a transaction and you don't start a new one.

Also you could add an early return in here when `!_enableAutoTransactions` what `_startTransaction` already does.

File: flutter/lib/src/navigation/time_to_display_tracker.dart [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+
+import '../../sentry_flutter.dart';
+import 'time_to_initial_display_tracker.dart';
+
+@internal
+class TimeToDisplayTracker {
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/time_to_display_tracker.dart:11]: `l`: This class is a bit obsolete right now, as it is a wrapper around `TimeToInitialDisplayTracker`.

File: flutter/lib/src/navigation/time_to_initial_display_tracker.dart [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+
+    final tracer = transaction as SentryTracer;
+
+    final ttidSpan = transaction.startChild(
+      SentrySpanOperations.uiTimeToInitialDisplay,
+      description: '${tracer.name} initial display',
+      startTimestamp: startTimestamp,
+    );
+
+    // Set the origin based on provided value or determine based on _isManual
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/time_to_initial_display_tracker.dart]: `l`: You can ditch this comment. For me, the code is self explaining.

File: flutter/lib/src/navigation/time_to_initial_display_tracker.dart [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+  }
+
+  Future<void> _trackTimeToInitialDisplay({
+    required ISentrySpan transaction,
+    required DateTime startTimestamp,
+    DateTime? endTimestamp,
+    String? origin,
+  }) async {
+    // Determine endTimestamp if not provided
+    final _endTimestamp = endTimestamp ?? await determineEndTime();
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/time_to_initial_display_tracker.dart:68]: `h`: Shouldn't we add some timeout for this? Otherwise, this could last forever.

File: flutter/lib/src/navigation/sentry_display_widget.dart [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+  @override
+  _SentryDisplayWidgetState createState() => _SentryDisplayWidgetState();
+}
+
+class _SentryDisplayWidgetState extends State<SentryDisplayWidget> {
+  @override
+  void initState() {
+    super.initState();
+    TimeToInitialDisplayTracker().markAsManual();
```
Comment (philipphofmann) [inline flutter/lib/src/navigation/sentry_display_widget.dart:49]: `h`: So when using this widget once, everything is in manual mode. I think instead it would be good to find a way to be able to use both combined.

File: flutter/test/navigation/sentry_display_widget_test.dart [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+    final ttidSpan = spans.first;
+    expect(ttidSpan.context.operation,
+        SentrySpanOperations.uiTimeToInitialDisplay);
+    expect(ttidSpan.finished, isTrue);
+    expect(ttidSpan.context.description, 'Current Route initial display');
+    expect(ttidSpan.origin, SentryTraceOrigins.manualUiTimeToDisplay);
+    expect(tracer.measurements, hasLength(1));
+    final measurement = tracer.measurements['time_to_initial_display'];
+    expect(measurement, isNotNull);
+    expect(measurement?.unit, DurationSentryMeasurementUnit.milliSecond);
```
Comment (philipphofmann) [inline flutter/test/navigation/sentry_display_widget_test.dart:53]: `m`: Shouldn't we also add an assert for the value of the measurement?

File: flutter/test/navigation/sentry_display_widget_test.dart [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+        element.context.operation ==
+        SentrySpanOperations.uiTimeToInitialDisplay);
+
+    expect(spans, hasLength(1));
+    final ttidSpan = spans.first;
+    expect(ttidSpan.context.operation,
+        SentrySpanOperations.uiTimeToInitialDisplay);
+    expect(ttidSpan.finished, isTrue);
+    expect(ttidSpan.context.description, 'Current Route initial display');
+    expect(ttidSpan.origin, SentryTraceOrigins.manualUiTimeToDisplay);
```
Comment (philipphofmann) [inline flutter/test/navigation/sentry_display_widget_test.dart:46]: `m`: I think adding asserts for the duration of the ttidSpan would improve this test.

File: flutter/test/navigation/time_to_initial_display_tracker_test.dart [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+
+      final children = transaction.children;
+      expect(children, hasLength(1));
+
+      final ttidSpan = children.first;
+      expect(ttidSpan.context.operation,
+          SentrySpanOperations.uiTimeToInitialDisplay);
+      expect(ttidSpan.finished, isTrue);
+      expect(ttidSpan.context.description, 'root ("/") initial display');
+      expect(ttidSpan.origin, SentryTraceOrigins.autoUiTimeToDisplay);
```
Comment (philipphofmann) [inline flutter/test/navigation/time_to_initial_display_tracker_test.dart:44]: `m`: Again I think we should assert the duration of the span.

**Review by philipphofmann — COMMENTED** [review]
As already discussed, my Flutter knowledge is still very limited, and therefore, I'm not confident enough to approve this PR. I added some comments.

File: flutter/lib/src/navigation/sentry_display_widget.dart (hunk 18 above)
Comment (buenaflor) [reply flutter/lib/src/navigation/sentry_display_widget.dart:49]: You can use both combined. You just don't wrap the screens you don't want to manually instrument

File: flutter/test/navigation/sentry_display_widget_test.dart (hunk 19 above)
Comment (buenaflor) [reply flutter/test/navigation/sentry_display_widget_test.dart:53]: It's kind of hard to reliably do it in this test

```
      fixture.navigatorObserver.didPush(currentRoute, null);
      await tester.pumpWidget(fixture.getSut());
```

we don't know how long `pumpWidget` is going to take until it finishes, so it's really in the hands of the testing framework

I guess we could add a threshold? something like if it takes longer than 300ms the test fails

File: flutter/test/navigation/sentry_display_widget_test.dart (hunk 20 above)
Comment (buenaflor) [reply flutter/test/navigation/sentry_display_widget_test.dart:46]: Same comment as above

File: flutter/lib/src/navigation/time_to_display_tracker.dart (hunk 15 above)
Comment (buenaflor) [reply flutter/lib/src/navigation/time_to_display_tracker.dart:11]: TTFD is also going to be added to this class, not sure if I should refactor it now

File: flutter/lib/src/navigation/sentry_navigator_observer.dart (hunk 14 above)
Comment (buenaflor) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart:268]: I think I'd rather rename `_finishTransaction`, `_startTransaction` is already a rather big function, combining all of them would make it a pretty big function

File: flutter/lib/src/navigation/sentry_display_widget.dart (hunk 18 above)
Comment (philipphofmann) [reply flutter/lib/src/navigation/sentry_display_widget.dart:49]: Then ignore me. Exellent 💯

File: flutter/test/navigation/sentry_display_widget_test.dart (hunk 19 above)
Comment (philipphofmann) [reply flutter/test/navigation/sentry_display_widget_test.dart:53]: The value should match the value of the span. Can you use that?

File: flutter/lib/src/navigation/time_to_display_tracker.dart (hunk 15 above)
Comment (philipphofmann) [reply flutter/lib/src/navigation/time_to_display_tracker.dart:11]: Then keep it. We can also ditch it later.

File: flutter/lib/src/navigation/sentry_navigator_observer.dart (hunk 14 above)
Comment (philipphofmann) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart:268]: I don't mind long functions if they clearly communicate what's going on. If the logic is simple and long, I don't mind having it in one function. Up to you.

File: flutter/test/navigation/sentry_display_widget_test.dart (hunk 19 above)
Comment (buenaflor) [reply flutter/test/navigation/sentry_display_widget_test.dart:53]: Added 👍

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 22]
```diff
@@ -88,13 +97,25 @@ class SentryNavigatorObserver extends RouteObserver<PageRoute<dynamic>> {
   final AdditionalInfoExtractor? _additionalInfoProvider;
   final SentryNative? _native;
 
+  static TimeToDisplayTracker? _timeToDisplayTracker;
+
+  @internal
+  static TimeToDisplayTracker? get timeToDisplayTracker =>
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]: this getter is not used anywhere, so we can remove it

File: flutter/lib/src/navigation/sentry_navigator_observer.dart [hunk 23]
```diff
@@ -88,13 +97,25 @@ class SentryNavigatorObserver extends RouteObserver<PageRoute<dynamic>> {
   final AdditionalInfoExtractor? _additionalInfoProvider;
   final SentryNative? _native;
 
+  static TimeToDisplayTracker? _timeToDisplayTracker;
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/sentry_navigator_observer.dart]:
```suggestion
  late final TimeToDisplayTracker? _timeToDisplayTracker;
```

Comment (buenaflor) [reply flutter/lib/src/navigation/sentry_navigator_observer.dart]: refactored it to be initialized in the constructor's initializer list directly so we don't need `late`

File: flutter/test/navigation/time_to_display_tracker_test.dart [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+
+  group('time to initial display', () {
+    group('in root screen app start route', () {
+      test('matches startTimestamp of transaction', () async {
+        final sut = fixture.getSut();
+
+        await sut.trackRegularRouteTTD(fixture.getTransaction(name: '/'),
+            startTimestamp: fixture.startTimestamp);
+
+        final transaction = fixture.hub.getSpan() as SentryTracer;
```
Comment (stefanosiano) [inline flutter/test/navigation/time_to_display_tracker_test.dart]:
```suggestion
        final transaction = fixture.getTransaction(name: '/') as SentryTracer;
        await sut.trackRegularRouteTTD(transaction, startTimestamp: fixture.startTimestamp);
```
this way it's easier to understand what starts a transaction and what finishes it
apply to other tests as well

File: flutter/test/navigation/time_to_display_tracker_test.dart [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+                element.context.operation ==
+                SentrySpanOperations.uiTimeToInitialDisplay)
+            .first;
+
+        expect(transaction, isNotNull);
+        expect(transaction.context.operation, SentrySpanOperations.uiLoad);
+        expect(transaction.startTimestamp, ttidSpan.startTimestamp);
+      });
+
+      test('startMeasurement finishes ttid span', () async {
```
Comment (stefanosiano) [inline flutter/test/navigation/time_to_display_tracker_test.dart]:
```suggestion
      test('trackAppStartTTD finishes ttid span', () async {
```
guess that's what the test is for, right?
apply to other tests as well

File: flutter/test/navigation/time_to_display_tracker_test.dart [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+      });
+
+      group('with manual strategy', () {
+        test('finishes ttid span after reporting with manual api', () async {
+          final sut = fixture.getSut();
+
+          Future.delayed(const Duration(milliseconds: 100), () {
+            fixture.ttidTracker.markAsManual();
+            fixture.ttidTracker.completeTracking();
+          });
```
Comment (stefanosiano) [inline flutter/test/navigation/time_to_display_tracker_test.dart]:
```suggestion
          fixture.ttidTracker.markAsManual();
          Future.delayed(const Duration(milliseconds: 1), () {
            fixture.ttidTracker.completeTracking();
          });
```
all other delays in all tests of this class seems useless, as they pass even without delays

File: flutter/test/navigation/time_to_display_tracker_test.dart [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+                  element.context.operation ==
+                  SentrySpanOperations.uiTimeToInitialDisplay)
+              .first;
+          expect(ttidSpan, isNotNull);
+
+          await Future.delayed(const Duration(milliseconds: 100));
+
+          expect(ttidSpan.finished, isTrue);
+        });
+      });
```
Comment (stefanosiano) [inline flutter/test/navigation/time_to_display_tracker_test.dart:147]: You can add another test for the timeout
Also, we can reduce the default timeout of `FakeFrameCallbackHandler` from 500 to 50 ms, to speed up tests, if it doesn't create issues

File: flutter/lib/src/navigation/time_to_initial_display_tracker.dart [hunk 28]
```diff
@@ -0,0 +1,135 @@
...
+        unit: ttidMeasurement.unit);
+    await ttidSpan.finish(endTimestamp: _endTimestamp);
+  }
+
+  Future<DateTime?>? determineEndTime() {
+    _trackingCompleter = Completer<DateTime?>();
+
+    // If we already know it's manual we can return the future immediately
+    if (_isManual) {
+      return _trackingCompleter?.future;
```
Comment (stefanosiano) [inline flutter/lib/src/navigation/time_to_initial_display_tracker.dart]: we can put a timeout here, too

File: flutter/test/navigation/time_to_display_tracker_test.dart (hunk 26 above)
Comment (buenaflor) [reply flutter/test/navigation/time_to_display_tracker_test.dart]: oh yeah my bad, the delays are not needed anymore since await already waits for everything to finish, wasn't the case before

**Review by stefanosiano — APPROVED** [review]
looks good to me! @krystofwoldrich wanna have a look, too?

**Comment by stefanosiano** [comment]
what about merging `SentryDisplayWidget` into `SentryWidget`?
So that if `SentryWidget` is used in a screen (it shouldn't matter where in the hierarchy) we can collect the TTID on "manual" mode.
This is just a question to open a discussion, we would do it in another PR

**Comment by buenaflor** [comment]
`SentryWidget` is meant to be used only once at the root (need to document that properly) so I don't think it's viable currently to merge them.

Gotta check what happens if we use `SentryWidget` multiple times.

But we can explore that option for sure

**Comment by krystofwoldrich** [comment]
Looks good! 🚀 Very exciting, looking forward to TTFD.
